cluster/distributedtask/scheduler.go

```diff
@@ -0,0 +1,365 @@
+type Scheduler struct {
+	mu           sync.Mutex
+	runningTasks map[string]map[TaskDescriptor]TaskHandle
+
+	providers          map[string]Provider // namespace -> Provider
+	completionRecorder TaskCompletionRecorder
+	tasksLister        TasksLister
+	taskCleaner        TaskCleaner
+	clock              clockwork.Clock
+
+	localNode        string
+	completedTaskTTL time.Duration
+	tickInterval     time.Duration
+
+	logger        logrus.FieldLogger
+	sampledLogger *logrusext.Sampler
+
+	tasksRunning *prometheus.GaugeVec
+
+	stopCh chan struct{}
+}
+
+type SchedulerParams struct {
+	CompletionRecorder TaskCompletionRecorder
+	TasksLister        TasksLister
+	TaskCleaner        TaskCleaner
+	Providers          map[string]Provider
+	Clock              clockwork.Clock
+	Logger             logrus.FieldLogger
+	MetricsRegisterer  prometheus.Registerer
+
+	LocalNode        string
+	CompletedTaskTTL time.Duration
+	TickInterval     time.Duration
+}
+
+func NewScheduler(params SchedulerParams) *Scheduler {
+	if params.Clock == nil {
+		params.Clock = clockwork.NewRealClock()
+	}
+
+	if params.MetricsRegisterer == nil {
+		params.MetricsRegisterer = monitoring.NoopRegisterer
+	}
+
+	return &Scheduler{
+		runningTasks: map[string]map[TaskDescriptor]TaskHandle{},
+
+		providers:          params.Providers,
+		completionRecorder: params.CompletionRecorder,
+		tasksLister:        params.TasksLister,
+		taskCleaner:        params.TaskCleaner,
+		clock:              params.Clock,
+
+		localNode:        params.LocalNode,
+		completedTaskTTL: params.CompletedTaskTTL,
+		tickInterval:     params.TickInterval,
+
+		logger:        params.Logger,
+		sampledLogger: logrusext.NewSampler(params.Logger, 5, 5*params.TickInterval),
+
+		tasksRunning: promauto.With(params.MetricsRegisterer).NewGaugeVec(prometheus.GaugeOpts{
+			Name: "weaviate_distributed_tasks_running",
+			Help: "Number of active distributed tasks running per namespace",
+		}, []string{"namespace"}),
+
+		stopCh: make(chan struct{}),
+	}
+}
+
+func (s *Scheduler) Start(ctx context.Context) error {
+	tasksByNamespace, err := s.listTasks(ctx)
+	if err != nil {
+		return errors.Wrap(err, "list distributed tasks")
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	for namespace, provider := range s.providers {
+		provider.SetCompletionRecorder(s.completionRecorder)
+
+		var (
+			tasks         = tasksByNamespace[namespace]
+			startedTasks  = s.filterStartedTasks(tasks)
+			localTaskDesc = provider.GetLocalTasks()
+		)
+		for _, taskDesc := range localTaskDesc {
+			if _, ok := startedTasks[taskDesc]; ok {
+				continue
+			}
+
+			if err := provider.CleanupTask(taskDesc); err != nil {
+				s.logger.WithFields(logrus.Fields{
+					"namespace":   namespace,
+					"taskID":      taskDesc.ID,
+					"taskVersion": taskDesc.Version,
+					"error":       err,
+				}).Error("failed to clean up local distributed task state")
```

Comment:
I think that there's a `WithError` method that you could use to log the error

Reply:
refactored